weave/trace/vals.py

```diff
@@ -323,14 +339,54 @@ def set_prefetched_rows(self, prefetched_rows: list[dict]) -> None:
         self._prefetched_rows = prefetched_rows
 
     def __len__(self) -> int:
-        return len(self.rows)
+        # This should e a single query
+        if self._known_length is not None:
+            return self._known_length
+
+        # Condition 1: we already have all the rows in memory
+        if self._prefetched_rows is not None:
+            self._known_length = len(self._prefetched_rows)
+            return self._known_length
+
+        # Condition 2: we have the row digests and they are a list
+        if (
+            self.table_ref is not None
+            and self.table_ref._row_digests is not None
+            and isinstance(self.table_ref._row_digests, list)
+        ):
+            self._known_length = len(self.table_ref._row_digests)
+            return self._known_length
+
+        # Condition 2: We don't know the length, in which case we can get it from the server
+        if self.table_ref is not None:
+            self._known_length = self._fetch_remote_length()
+            return self._known_length
+
+        # Finally, if we have no table ref, we can still get the length
+        # by materializing the rows as a list. I actually think this
+        # can never happen, but it is here for completeness.
+        rows_as_list = self._inefficiently_materialize_rows_as_list()
+        return len(rows_as_list)
```

Comment:
shouldn't also set `self._known_length` here?

Reply:
we should set the length any time we `_inefficiently_materialize_rows_as_list`